jerry-core/ecma/operations/ecma-typedarray-object.c

```diff
@@ -0,0 +1,941 @@
+{
+  if (!ecma_is_value_object (target))
+  {
+    return false;
+  }
+
+  ecma_object_t *obj_p = ecma_get_object_from_value (target);
+
+  if (ecma_get_object_type (obj_p) == ECMA_OBJECT_TYPE_PSEUDO_ARRAY)
+  {
+    ecma_extended_object_t *ext_object_p = (ecma_extended_object_t *) obj_p;
+
+    return ((ext_object_p->u.pseudo_array.type == ECMA_PSEUDO_ARRAY_TYPEDARRAY)
+            || (ext_object_p->u.pseudo_array.type == ECMA_PSEUDO_ARRAY_TYPEDARRAY_WITH_INFO));
+  }
+
+  return false;
+} /* ecma_is_typedarray */
+
+/**
+ * List names of a TypedArray object's integer indexed properties
+ *
+ * @return void
+ */
+void
+ecma_op_typedarray_list_lazy_property_names (ecma_object_t *obj_p, /**< a TypedArray object */
+                                             ecma_collection_header_t *main_collection_p) /**< 'main'
+                                                                                           *   collection */
+{
+  JERRY_ASSERT (ecma_is_typedarray (ecma_make_object_value (obj_p)));
+
+  ecma_length_t array_length = ecma_typedarray_get_length (obj_p);
+
+  for (ecma_length_t i = 0; i < array_length; i++)
+  {
+    ecma_string_t *name_p = ecma_new_ecma_string_from_uint32 (i);
+
+    ecma_append_to_values_collection (main_collection_p, ecma_make_string_value (name_p), true);
+
+    ecma_deref_ecma_string (name_p);
+  }
+} /* ecma_op_typedarray_list_lazy_property_names */
+
+/**
+ * Get the integer number property value of the typedarray
+ *
+ * See also: ES2015 9.4.5.8
+ *
+ * @return ecma value
+ *         if failed, return undefined value
+ */
+ecma_value_t
+ecma_op_typedarray_get_index_prop (ecma_object_t *obj_p, /**< a TypedArray object */
+                                   uint32_t index) /**< the index number */
+{
+  JERRY_ASSERT (ecma_is_typedarray (ecma_make_object_value (obj_p)));
+
+  ecma_length_t array_length = ecma_typedarray_get_length (obj_p);
+
+  if (index >= array_length)
+  {
+    return ecma_make_simple_value (ECMA_SIMPLE_VALUE_UNDEFINED);
+  }
+
+  ecma_object_t *arraybuffer_p = ecma_typedarray_get_arraybuffer (obj_p);
+  ecma_length_t offset = ecma_typedarray_get_offset (obj_p);
+  uint8_t shift = ecma_typedarray_get_element_size_shift (obj_p);
+  ecma_length_t byte_pos = (index << shift) + offset;
+  lit_magic_string_id_t class_id = ecma_object_get_class_name (obj_p);
+  lit_utf8_byte_t *target_p = ecma_arraybuffer_get_buffer (arraybuffer_p) + byte_pos;
+  ecma_value_t value;
+
+#define TYPEDARRAY_GET_INDEXED_ELEMENT(type) \
+  type v = *((type *) target_p); \
+  value = ecma_make_number_value ((ecma_number_t) v);
+
+  switch (class_id)
+  {
+    case LIT_MAGIC_STRING_INT8_ARRAY_UL:
+    {
+      TYPEDARRAY_GET_INDEXED_ELEMENT (int8_t);
+      break;
+    }
+    case LIT_MAGIC_STRING_UINT8_ARRAY_UL:
+    case LIT_MAGIC_STRING_UINT8_CLAMPED_ARRAY_UL:
+    {
+      TYPEDARRAY_GET_INDEXED_ELEMENT (uint8_t);
+      break;
+    }
+    case LIT_MAGIC_STRING_INT16_ARRAY_UL:
```

Comment:
About the endian problem, 
I mean, native compiler will handle the Endian problem, and for us, it is only type-cast.
For example, we want to set 2 uint16_t number 0x1122 and 0x3344 in the Uint16Array, and convert to Uint8Array
```
var a = new Uint16Array(2);
a[0]=0x1122;
a[1]=0x3344;
var b = new Uint8Array(a); //[0x22, 0x44]
```
In little endian, the data in the arraybuffer of a, is (hex) 22 11 44 33
after the convertion-to-uint8, the data in the arraybuffer of b, is (hex) 22 44

In big endian, the data in the arraybuffer of a, is (hex) 11 22 33 44
after the convertion-to-uint8, the data in the arraybuffer of b, is (hex) 22 44

How uint16_t value (0x1122, and 0x3344) stored in the memory, is handled by compiler. And how to change a uint16_t value to uint8_t, is also handled by compiler.

So jerry dont need to deal with the endian problem
Am I correct about that?